File: Data Structures & Algorithms/binary-tree-diameter/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    public int maxDepth(TreeNode root) {
        if (root==null) return 0;
        return 1+Math.max(maxDepth(root.left),maxDepth(root.right));
    }
    public int diameterOfBinaryTree(TreeNode root) {
        if (root==null) return 0;

        int najvece_rastojanje = (maxDepth(root.left) + maxDepth(root.right));

        int najvece_rastojanje_levo = diameterOfBinaryTree(root.left);
        int najvece_rastojanje_desno = diameterOfBinaryTree(root.right);

        return Math.max(najvece_rastojanje,Math.max(najvece_rastojanje_desno,najvece_rastojanje_levo));
    }
}

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   


    def max_depth(self,p:TreeNode):
        if not p:return 0


        return 1+max(self.max_depth(p.left),self.max_depth(p.right))

    def isSameTree(self, p: TreeNode, q: TreeNode) -> bool:
        solucija = Solution()
        visina_p = solucija.max_depth(p)
        visina_q = solucija.max_depth(q)

        if visina_p!=visina_q: return False

        return self.isSameTree_helper(p,q)

    def isSameTree_helper(self, p: TreeNode, q: TreeNode) -> bool:

        if (not p) and (not q): return True

        if (p and not q) or (not p and q):return False

        if p.val==q.val:
            return self.isSameTree_helper(p.left,q.left) and self.isSameTree_helper(p.right,q.right)
        return False

    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root:return False

        ima_li_ga = self.isSameTree(root,subRoot)

        return ima_li_ga or self.isSubtree(root.left, subRoot) or self.isSubtree(root.right,subRoot)
